File: 算法/src/js/04_breadth_first_search.js
/**
 * 广度优先遍历
 */

const person_is_seller = name => name[name.length - 1] === 'm';

graph = {};
graph['you'] = ['alice', 'bob', 'claire'];
graph['bob'] = ['anuj', 'peggy'];
graph['alice'] = ['peggy'];
graph['claire'] = ['thom', 'jonny'];
graph['anuj'] = [];
graph['peggy'] = [];
graph['thom'] = [];
graph['jonny'] = [];

const breadth_first_search = name => {
  let search_queue = graph[name];
  let searched = [];
  while (search_queue.length) {
    let person = search_queue.shift();
    if (!searched.includes(person)) {
      if (person_is_seller(person)) {
        console.log(person);
        return true;
      } else {
        searched.push(person);
        search_queue = search_queue.concat(graph[person]);
      }
    }
  }
  return false;
};

breadth_first_search('you');

File: algorithm/src/python/04_breadth_first_search.py
# coding=utf-8

# 广度优先遍历
from collections import deque


def person_is_seller(name):
    print(name)
    return name[-1] == "m"


graph = {}
graph["you"] = ["alice", "bob", "claire"]
graph["bob"] = ["anuj", "peggy"]
graph["alice"] = ["peggy"]
graph["claire"] = ["thom", "jonny"]
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []


def breath_first_search(name):
    search_queue = deque()
    search_queue += graph[name]

    searched = []
    while(search_queue):
        person = search_queue.popleft()

        if person not in searched:
            if person_is_seller(person):
                print(person + " is a seller.")
                return True
            else:
                search_queue += graph[person]
                searched.append(person)
    return False


breath_first_search("you")
